### PR TITLE
Fix Redundant ScalarValue Boxed Collection

### LINKED ISSUE #2449: Redundant ScalarValue Boxed Collection
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

`ScalarValue::List` and `ScalarValue::Struct` contain a `Box<Vec<_>>` this adds a redundant level of boxing. Clippy in fact complains about this, but this has been suppressed with `#[allow(clippy::box_collection)]`

**Describe the solution you'd like**

Remove the additional `Box`

### PR DESCRIPTION
**Pull request by comphead**
# Which issue does this PR close?

Closes #2449.

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -74,8 +74,7 @@ pub enum ScalarValue {
     /// large binary
     LargeBinary(Option<Vec<u8>>),
     /// list of nested ScalarValue (boxed to reduce size_of(ScalarValue))
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:76]:
```suggestion
    /// list of nested ScalarValue
```

File: datafusion/common/src/scalar.rs [hunk 2]
```diff
@@ -95,8 +94,7 @@ pub enum ScalarValue {
     /// Interval with MonthDayNano unit
     IntervalMonthDayNano(Option<i128>),
     /// struct of nested ScalarValue (boxed to reduce size_of(ScalarValue))
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:96]:
```suggestion
    /// struct of nested ScalarValue
```

**Review by alamb — COMMENTED** [review]
Thank you very much @comphead 

They key test is that `size_of_scalar` is not changed

https://github.com/apache/arrow-datafusion/blob/41b4e491663029f653e491b110d0b5e74d08a0b6/datafusion/core/src/scalar.rs#L373-L383

However, when I investigated more it turns out that this test was broken (see https://github.com/apache/arrow-datafusion/pull/2531)

When I reenabled the test it shows that this PR makes the size of `ScalarValue` increase from `48` bytes to `64` bytes 

cc @tustvold

File: datafusion/common/src/scalar.rs [hunk 3]
```diff
@@ -74,8 +74,7 @@ pub enum ScalarValue {
     /// large binary
     LargeBinary(Option<Vec<u8>>),
     /// list of nested ScalarValue (boxed to reduce size_of(ScalarValue))
-    #[allow(clippy::box_collection)]
-    List(Option<Box<Vec<ScalarValue>>>, Box<DataType>),
+    List(Option<Vec<ScalarValue>>, DataType),
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:77]: I'm not sure about unboxing DataType here

**Review by tustvold — COMMENTED** [review]
I think this PR removes a non-redundant `Box<DataType>`. Is it possible this is the cause of the size change?

**Comment by tustvold** [comment]
So I got this to not change the size by making

```
List(Option<Vec<ScalarValue>>, Box<DataType>),
Struct(Option<Vec<ScalarValue>>, Box<[Field]>),
```

@alamb What do you think about `Box<[Field]>`?

**Comment by comphead** [comment]
Thanks @alamb @tustvold for the feedback, please let me know the final decision.

**Comment by alamb** [comment]
> @alamb What do you think about Box<[Field]>?

It seems fine to me

**Comment by comphead** [comment]
@tustvold I added boxing back.
For now its 
```
Struct(Option<Vec<ScalarValue>>, Box<Vec<Field>>),
```
as its more consistent, you are right. If you feel we should go to Box<[Field]> let me know

File: datafusion/common/src/scalar.rs [hunk 4]
```diff
@@ -934,16 +928,20 @@ impl ScalarValue {
                         match values {
                             Some(values) => {
                                 // Push value for each field
-                                for c in 0..columns.len() {
-                                    let column = columns.get_mut(c).unwrap();
-                                    column.push(values[c].clone());
+                                for (i, v) in
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:931]: 👍  this is a nicer implementation

**Review by alamb — APPROVED** [review]
I think this looks great @comphead  -- thank you very much for sticking with it. 

I also merged this PR locally with `apache/master` to get https://github.com/apache/arrow-datafusion/pull/2531 and verified that the tests still pass 🎉

**Comment by alamb** [comment]
I'll wait until tomorrow to see if @tustvold  has any more comments, and if not I'll merge it in

File: datafusion/common/src/scalar.rs (hunk 4 above)
Comment (comphead) [reply datafusion/common/src/scalar.rs:931]: clippy complained on that :)

Comment (alamb) [reply datafusion/common/src/scalar.rs:931]: All the better that removing a layer of `Box`ing lets clippy fix things more easily 👍

Comment (tustvold) [reply datafusion/common/src/scalar.rs:931]: A perhaps more idiomatic way to write this would be something like

```
for (column, value) in columns.iter_mut().zip(values)
```

File: datafusion/core/src/scalar.rs [hunk 5]
```diff
@@ -643,11 +643,11 @@ mod tests {
         // For different data type, `partial_cmp` returns None.
         assert_eq!(
             List(
-                Some(Box::new(vec![Int64(Some(1)), Int64(Some(5))])),
-                Box::new(DataType::Int64),
+                Some(vec![Int64(Some(1)), Int64(Some(5))]),
+                Box::new(DataType::Int32),
```
Comment (tustvold) [inline datafusion/core/src/scalar.rs:647]: Why the change in type?

Comment (comphead) [reply datafusion/core/src/scalar.rs]: fixed. Thanks for finding that

File: datafusion/common/src/scalar.rs (hunk 4 above)
Comment (comphead) [reply datafusion/common/src/scalar.rs]: Yes, thats more concise. Fixed in 2 places.

**Comment by alamb** [comment]
image
👍
